### ISSUE TITLE
trainer.py:163: RuntimeWarning: invalid value encountered in true_divide

### ISSUE THREAD

**Issue by MDReptile** [Unity-Technologies/ml-agents#230]
Forgive me ahead of time for this huge mess of prototype code that won't format properly here (probably lots of problems)! I've chopped the code down to what I think is relevant here, which is the agent script and the controller script where it rewards and punishes the agent.

**Sometimes I'm getting this error message in the jupyter notebook PPO, as I'm training, and then the agents stop working (no longer giving input)**:

`c:\UnityProjects\DeepBrainz\python\ppo\trainer.py:163: RuntimeWarning: invalid value encountered in true_divide
  self.training_buffer['advantages'] = (advantages - advantages.mean()) / advantages.std()`

Here is how I have this agent set up:

_____________________________________________

 public class BrainzAgent : Agent
 {
    private CharacterControls cControls;

    public bool actionLeft = false, actionRight = false, 
        actionForward = false, actionBack = false, actionAttack = false;

    public GameObject ActiveCharacter;

    public override void InitializeAgent()
    {
        if(ActiveCharacter != null)
            cControls = ActiveCharacter.GetComponent<CharacterControls>();
    }

    List<float> state = new List<float>();

    // one-hots
    public bool RayL1Hit, ... RayR5Hit;
    public float RayL1Dist, ... RayR5Dist;

    public override List<float> CollectState()
    {
        // 13 + 11 = 24 total state vars
        state.Clear();

        if (cControls == null || ActiveCharacter == null)
        {
            // blank info, till respawned
            for(int i = 0; i < 24; i++)
                state.Add(0);

            if (ActiveCharacter != null)
            {
                cControls = ActiveCharacter.GetComponent<CharacterControls>();
                done = true; // cause reset
            }
        }
        else
        {
            if (cControls.IsHuman) { state.Add(1); }
            else { state.Add(0); }
            if (cControls.HasWeapon) { state.Add(1); }
            else { state.Add(0); }
            if (RayL1Hit) { state.Add(1); }
            else { state.Add(0); }
           // ..... cut for space
            if (RayR5Hit) { state.Add(1); }
            else { state.Add(0); }

            state.Add(RayL1Dist);
            // ..... cut for space
            state.Add(RayR5Dist);
        }

        return state;
    }

    public override void AgentStep(float[] action)
    {
        switch ((int)action[0])
        {
            case 0: // do nothing
                actionRight = false;
                actionLeft = false;
                actionForward = false;
                actionBack = false;
                actionAttack = false;
                break;
             //    .. unrelated variants of above and below
            case 5: // attack
                actionRight = false;
                actionLeft = false;
                actionForward = false;
                actionBack = false;
                actionAttack = true;
                break;
            default:
                return;
        }
    }

    public void Fail(bool zombieDeath)
    {
        //Debug.Log("Fail");
        reward -= 1f; // failure
        if(zombieDeath) // character is completely dead
            ActiveCharacter = null;
        done = true; // reset will be called
    }

    public void GiveReward(int points)
    {
        //Debug.Log("success");
        reward += 0.1f * points; // success
    }

    public override void AgentReset()
    {
        // called after done becomes true
        //Debug.Log("Reset Agent");
    }
 }

_____________________________________________
and the rewards are given through this script for controlling the simple characters:
_____________________________________________

[RequireComponent(typeof(Rigidbody))]
public class CharacterControls : MonoBehaviour
{
    public GameObject WeaponPrefab, BulletPrefab;
    public Transform FirePos;
    public bool PlayerControlled = false;
    public bool HasWeapon = false;
    public bool IsHuman = true;
    private BrainzAgent agent;

    void FixedUpdate()
    {
        if(agent != null)// AI controlled
        {
            // RAYCASTING ----------------------

            // INPUT ----------------------------
            // use input from agent
            if (agent.actionForward)
                rb.AddForce(transform.forward * movePower);
            else if (agent.actionBack)
                rb.AddForce(-transform.forward * movePower);

            if (agent.actionLeft)
                transform.Rotate(new Vector3(0, -turnRate, 0));
            else if (agent.actionRight)
                transform.Rotate(new Vector3(0, turnRate, 0));

            if (IsHuman && HasWeapon)
            {
                if (agent.actionAttack)
                {
                    AttemptFireWeapon();
                }
            }
            else if (!IsHuman) // player controlled zombie
            {
                if (agent.actionAttack)
                {
                    AttemptZombieMelee();
                }
            }
        }

        if (keepUpright)
            KeepUpright();
    }

    public void HitEnemy()
    {
        // hit an enemy with bullet
        if(!PlayerControlled)
            agent.GiveReward(10);
    }

    public void SetAgent(BrainzAgent a) // suspended agent is assigned to newly created character
    {
        agent = a;
    }

    float lastFireTime, fireDelay = 0.5f;

    void AttemptFireWeapon()
    {
        if (!HasWeapon)
            Debug.LogError("Call AttemptFireWeapon with HasWeapon false!");

        if (Time.time - lastFireTime > fireDelay) // if enough time has passed since last shot
        {
            GetOrCreatePoolBullet(transform.forward);
            lastFireTime = Time.time;
        }
    }

    void AttemptZombieMelee()
    {
        if (!IsHuman && Time.time - lastFireTime > fireDelay) // if enough time has passed since last attack
        {
            MeleeAttack();
            lastFireTime = Time.time;
        }
    }

    void MeleeAttack()
    {
        Ray shotRay = new Ray(FirePos.position, transform.forward);
        RaycastHit hitInfo = new RaycastHit();
        if (Physics.Raycast(shotRay, out hitInfo, 1.5f))
        {
            if (hitInfo.collider.CompareTag("Human"))
            {
                // hit a human
                CharacterControls humanHit = hitInfo.collider.gameObject.GetComponent<CharacterControls>();
                // damage or kill the human
                humanHit.TakeHit(Random.Range(1, 3)); // take 1 or 2 damage (3 exclusive)
                agent.GiveReward(10);
            }
        }
    }

    public void TakeHit(int damage)
    {
        lifeLeft -= damage;
        if (lifeLeft <= 0)
        {
            // dead character
            Die();
        }
    }

    void Die()
    {
        // if human become zombie, if zombie, die completely
        if (IsHuman)
            GameMaster.Instance.CharactersH.Remove(this.gameObject);
        else
            GameMaster.Instance.CharactersZ.Remove(this.gameObject);

        if (IsHuman)
        {
            RemovePoolBullets();
            if (PlayerControlled)
            {
                BecomeZombie();
            }
            else
            {
                agent.Fail(false);
                BecomeZombie();
            }
            GameMaster.Instance.CharactersZ.Add(this.gameObject);
        }
        else // already zombie
        {
            if (!PlayerControlled)
                agent.Fail(true);
            keepUpright = false;
            Splode(); // kill the character and suspend the agent temporarily
        }
    }
    // ... non-related code
}

_____________________________________________

Any ideas what is wrong? I haven't modified the PPO really, just added a name and set the training steps.

**Comment by MDReptile** [comment]
I'm not totally sure, but I think this might have been because too many agents were getting 0 overall reward for entire runs, and it created a divide by zero. I'm guessing anyway, since I've changed rewarding around to make sure agents get either positive or negative rewards in more cases.

**Comment by kwea123** [comment]
Advantage value is actually `Q(s, a) - V(s)`, so unless your value function always outputs zero, even if the rewards are all zeros, it is not going to be zero. The problem is what you're dividing is its std, so if all the rewards are zeros and they evaluate on the same state, it could be zero. However this means your agent doesn't move at all...

I will suggest that you print out the advantage value to see what happened..., also personally I didn't normalize the advantage and the agent still learnt very well.

**Comment by MDReptile** [comment]
How can I make sure the std doesn't result in 0? Just give a minor reward every so often for being alive?

**Comment by kwea123** [comment]
Yes but I'm more worried about the fact that it is zero, because theoretically it shouldn't if your agent runs correctly.

I can only think of two ways that it's zero:
1. The advantage is evaluated on the same state, and the discounted reward is the same, which means your agent doesn't move.
2. There are only one value in the buffer, which means your agent is done and resets after one frame.

I think you need to check out why if it's either of these cases first

**Comment by MDReptile** [comment]
I have a "dynamic agent setup" where I'm testing 10 agents in the scene, and the game spawns 10 random characters and assigns each to a agent.

The characters are on two teams, and randomly the kill some of the agents characters, so the agent becomes "inactive" waiting for the round to end to have the game spawn more characters.

Perhaps this time where I don't do anything with input, and give all 0 floats as state, causes an issue? You can see this in BrainzAgent.cs near the top of FixedUpdate() that if the agent doesn't have a character assigned it gets all 0 floats.

Not a good idea?

**Comment by kwea123** [comment]
Well if the agent becomes inactive just don't train on that agent?

**Comment by MDReptile** [comment]
How do I stop the training temporarily on them?

**Comment by MDReptile** [comment]
Yeah I still hit this warning (which stops my agents from doing anything) every so often.

Wish I understood what I could do to stop it :/

**Comment by kwea123** [comment]
Yeah if the character is dead, then

```
if (cControls == null || ActiveCharacter == null)
    {
        // blank info, till respawned
        for(int i = 0; i < 24; i++)
            state.Add(0);
```

The states are all zeros until it respawns, you need to stop training on this agent! Otherwise it's the situation that I mentioned
> The advantage is evaluated on the same state, and the discounted reward is the same, which means your agent doesn't move.

Also I doubt if the agents will learn correctly, since this is a dynamic environment. Basically because agents try to do their best in response to their enemies (and their teammates), so the strategies interact with each other. PPO doesn't guarantee any convergence. You need to use algorithms that allow training on multiple agents (multi-agent reinforcement learning, MARL). You can find some information here https://github.com/LantaoYu/MARL-Papers

And as far as I know, MADDPG https://arxiv.org/pdf/1706.02275.pdf should be the only one currently having a clearly written pseudo code

**Comment by MDReptile** [comment]
Ahh yes, it must be because I was returning all those zeroes it looks like. At least I've been able to run it for a couple hundred thousand steps and not hit that warning after returning -1 instead. I'm happy to see I haven't broken something else.

I'll have to do some more research about ML in general and the stuff you shared to see if I can come up with an alternative way to train them, but I might be getting in a little over my head writing an alternative to the PPO that comes with the examples. Thanks for all the help with everything @kwea123 !

**Comment by vladimiroster** [comment]
Thanks for reaching out to us. Hopefully you were able to resolve your issue.  We are closing this due to inactivity.
